include/deal.II/base/tensor.h

```diff
@@ -300,153 +558,384 @@ class Tensor
                   << arg1
                   << ", but this is not possible for tensors of the current type.");
 
+private:
   /**
-   * Read or write the data of this object to or from a stream for the purpose
-   * of serialization
+   * Internal type declaration that is used to specialize the return type
+   * of operator[]() for Tensor<1,dim,Number>
    */
-  template <class Archive>
-  void serialize(Archive &ar, const unsigned int version);
+  typedef Tensor<rank_, dim, Number> tensor_type;
 
-private:
   /**
    * Array of tensors holding the subelements.
    */
-  Tensor<rank_-1,dim,Number> subtensor[dim];
+  Tensor<rank_-1, dim, Number> values[(dim != 0) ? dim : 1];
+  // ... avoid a compiler warning in case of dim == 0 and ensure that the
+  // array always has positive size.
 
   /**
    * Help function for unroll.
    */
-  template <typename Number2>
-  void unroll_recursion(Vector<Number2> &result,
-                        unsigned int    &start_index) const;
+  template <typename OtherNumber>
+  void unroll_recursion(Vector<OtherNumber> &result,
+                        unsigned int        &start_index) const;
 
-  // make the following class a
-  // friend to this class. in principle,
-  // it would suffice if otherrank==rank+1,
-  // but then the compiler complains
-  // that this be an explicit specialization
-  // which is not what we want
-  //
-  // also, it would be sufficient to make
-  // the function unroll_loops a friend,
-  // but that seems to be impossible as well.
+  /**
+   * Allow an arbitrary Tensor to access the underlying values.
+   */
   template <int, int, typename> friend class Tensor;
+
+  /**
+   * Point is allowed access to the coordinates. This is supposed to improve
+   * speed.
+   */
+  friend class Point<dim,Number>;
 };
 
 
-/*--------------------------- Inline functions -----------------------------*/
 
 #ifndef DOXYGEN
+/*---------------------- Inline functions: Tensor<0,dim> ---------------------*/
 
-template <int rank_, int dim, typename Number>
+
+template <int dim,typename Number>
 inline
-Tensor<rank_,dim,Number>::Tensor ()
+Tensor<0,dim,Number>::Tensor ()
 {
-// default constructor. not specifying an initializer list calls
-// the default constructor of the subobjects, which initialize them
-// selves. therefore, the tensor is set to zero this way
+  value = value_type();
```

Comment:
why not move this into a member initializer list as usual? same for the constructors below.